pkg/skaffold/build/cache.go

```diff
@@ -0,0 +1,368 @@
+/*
+Copyright 2019 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package build
+
+import (
+	"bytes"
+	"context"
+	"crypto/md5"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd/config"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/color"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/constants"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	homedir "github.com/mitchellh/go-homedir"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// ImageDetails holds the Digest and ID of an image
+type ImageDetails struct {
+	Digest string `yaml:"digest,omitempty"`
+	ID     string `yaml:"id,omitempty"`
+}
+
+// ArtifactCache is a map of [artifact dependencies hash : ImageDetails]
+type ArtifactCache map[string]ImageDetails
+
+// Cache holds any data necessary for accessing the cache
+type Cache struct {
+	artifactCache ArtifactCache
+	client        docker.LocalDaemon
+	cacheFile     string
+	useCache      bool
+}
+
+var (
+	// For testing
+	hashForArtifact = getHashForArtifact
+	localCluster    = config.GetLocalCluster
+)
+
+// NewCache returns the current state of the cache
+func NewCache(useCache bool, cacheFile string) *Cache {
```

Comment:
this is never called with anything else than `cacheFile == ""` in production code, only in test code. Did you forget to implement an extra flag to define the cacheFile?

Reply:
whoops, fixed. thanks!